PublicMapping/district-builder-2
# Add JWT support for auth on server

## Overview

Add JWT support for auth on server.

Verifying email address and logging in now return a JWT which is then required to access any endpoint with the `@UseGuards(JwtAuthGuard)` decorator.

### Demo

```
09:36 $ http http://localhost:3005/auth/email/register email=pc@azavea.com password=test name="Peter C"
HTTP/1.1 201 Created
Connection: keep-alive
Content-Length: 7
Content-Type: text/html; charset=utf-8
Date: Thu, 09 Apr 2020 13:49:45 GMT
ETag: W/"7-gFw9aUPkrb4JBgwx0K+SLEMaCno"
X-Powered-By: Express

SUCCESS

09:49 $ http POST http://localhost:3005/auth/email/resend-verification/peterc@azavea.com
HTTP/1.1 201 Created
Connection: keep-alive
Content-Length: 7
Content-Type: text/html; charset=utf-8
Date: Thu, 09 Apr 2020 13:49:51 GMT
ETag: W/"7-gFw9aUPkrb4JBgwx0K+SLEMaCno"
X-Powered-By: Express

SUCCESS

09:49 $ http POST http://localhost:3005/auth/email/verify/bEyUZJ5EvmGwolj_6CfBNl_kjSk
HTTP/1.1 201 Created
Connection: keep-alive
Content-Length: 337
Content-Type: text/html; charset=utf-8
Date: Thu, 09 Apr 2020 13:50:11 GMT
ETag: W/"151-7lVQ0pAulid4bQvlMsAftSFuR9k"
X-Powered-By: Express

eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJiNzZlYTk4OS1jYThjLTQyZWYtYjk4My00MWJmZTg4ZjAwMTIiLCJpZCI6ImI3NmVhOTg5LWNhOGMtNDJlZi1iOTgzLTQxYmZlODhmMDAxMiIsIm5hbWUiOiJQZXRlciBDIiwiZW1haWwiOiJwZXRlcmNAYXphdmVhLmNvbSIsImlzRW1haWxWZXJpZmllZCI6dHJ1ZSwiaWF0IjoxNTg2NDQwMjExLCJleHAiOjE1ODY2MTMwMTF9.S70DAhMZvnK_bTHo4u55cdhshhP5VrhdiIK_juwKoto

09:50 $ http http://localhost:3005/auth/email/login email=peterc@azavea.com password=test
HTTP/1.1 201 Created
Connection: keep-alive
Content-Length: 337
Content-Type: text/html; charset=utf-8
Date: Thu, 09 Apr 2020 13:53:18 GMT
ETag: W/"151-HVHdlcX+aTDK7F+3E06F7idk1QE"
X-Powered-By: Express

eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJiNzZlYTk4OS1jYThjLTQyZWYtYjk4My00MWJmZTg4ZjAwMTIiLCJpZCI6ImI3NmVhOTg5LWNhOGMtNDJlZi1iOTgzLTQxYmZlODhmMDAxMiIsIm5hbWUiOiJQZXRlciBDIiwiZW1haWwiOiJwZXRlcmNAYXphdmVhLmNvbSIsImlzRW1haWxWZXJpZmllZCI6dHJ1ZSwiaWF0IjoxNTg2NDQwMzk4LCJleHAiOjE1ODY2MTMxOTh9.iarmOv47Tm4u1bk2E-vRxKuqlAT6JStxnhhypsFLux8

11:49 $ http http://localhost:3005/users 'Authorization:Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJiNzZlYTk4OS1jYThjLTQyZWYtYjk4My00MWJmZTg4ZjAwMTIiLCJpZCI6ImI3NmVhOTg5LWNhOGMtNDJlZi1iOTgzLTQxYmZlODhmMDAxMiIsIm5hbWUiOiJQZXRlciBDIiwiZW1haWwiOiJwZXRlcmNAYXphdmVhLmNvbSIsImlzRW1haWxWZXJpZmllZCI6dHJ1ZSwiaWF0IjoxNTg2NDQ0NDM3LCJleHAiOjE1ODY2MTcyMzd9.EQkoDGMhA35rxm07OVBQtGOTBcdUrrfaw30XEuiWDWQ'
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 85
Content-Type: application/json; charset=utf-8
Date: Fri, 10 Apr 2020 15:49:33 GMT
ETag: W/"55-9Ga/LQZ+0gPa0ku5RMDlnWyuQ50"
X-Powered-By: Express

{
    "id": "b76ea989-ca8c-42ef-b983-41bfe88f0012",
    "isEmailVerified": true,
    "name": "Peter C"
}

11:49 $ http http://localhost:3005/users
HTTP/1.1 401 Unauthorized
Connection: keep-alive
Content-Length: 41
Content-Type: application/json; charset=utf-8
Date: Fri, 10 Apr 2020 16:07:06 GMT
ETag: W/"29-O6gP33ALmcYdbLhsI8ncjapAJ44"
X-Powered-By: Express

{
    "error": "Unauthorized",
    "statusCode": 401
}
```

### Notes

This implementation was largely based on the official docs: https://docs.nestjs.com/techniques/authentication

It includes a few refactors:
1. Remove the generic `ERROR` value from the error enums in `constants.ts`. Those generic errors weren't actually being returned by any function or consumed in any way except in the messages of exceptions at the very end. It seems like we should only have our services return _specific_ errors since if there's an unexpected error, the function will throw, so there doesn't seem to be much of a need to encode generic errors in types.
1. Relatedly, ddcbdf5 tries to bring uniformity to the way we're handling errors in the auth controller (see that commit for details). Essentially, because any service function could throw -- and it seems these 3rd party libraries often do -- the idea is to wrap the happy path in a `try` block and then inspect the error in the `catch`. For example:
```ts
    try {
      // this could throw (unexpected exception)
      const verifiedUser = await this.authService.verifyEmail(token);
      if (verifiedUser === undefined) {
        // here we intentionally throw to return a 404
        throw new NotFoundException(
          VerifyEmailErrors[VerifyEmailErrors.NOT_FOUND],
          "Email or user not found for token"
        );
      }
      // this could also throw
      return this.authService.generateJwt(verifiedUser);
    } catch (error) {
      if (error instanceof HttpException) {
        throw error;
      } else {
        this.logger.error(`Error verifying email token: ${error}`);
        throw new InternalServerErrorException();
      }
    }
```
The nice thing about using classes for exceptions is that we can use `instanceof` to determine what type of exception it is.

The motivation for this was errors cropping up that weren't caught by the type system (eg. if you fail to pass a `sub` into `JwtService.sign`, it will throw unexpectedly) since many of these functions take imprecise types and TS doesn't have any way to account for the fact that a function could throw. That combined with the fact that NestJS has you throw errors to return the appropriate HTTP status code made it hard to work with as it was before.

This seems like a pretty safe approach in that we'll be sure to catch errors, log appropriately, and not bubble up sensitive data, but it is boilerplate-y (perhaps we should discuss more how to better handle it).

## Testing Instructions
I think due to all the changes it's worth going through the full auth flow for testing as shown in the demo:

* Register a new user
* Re-request verification email
* Use the verification code to get a JWT
* Log in with that user (should get the same JWT back)
* Use that JWT to list users
* Make that same request without the JWT and ensure you get a 401

Closes #50
